File: src/examples/java/org/nodex/examples/composition/CompositionExample.java
/*
 * Copyright 2002-2011 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of the
 * License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package org.nodex.examples.composition;

import org.nodex.core.CompletionHandler;
import org.nodex.core.EventHandler;
import org.nodex.core.buffer.Buffer;
import org.nodex.core.composition.Composable;
import org.nodex.core.composition.Composer;
import org.nodex.core.file.FileSystem;
import org.nodex.core.http.HttpServer;
import org.nodex.core.http.HttpServerRequest;
import org.nodex.mods.amqp.AmqpClient;
import org.nodex.mods.amqp.AmqpConnectHandler;
import org.nodex.mods.amqp.AmqpConnection;
import org.nodex.mods.amqp.AmqpMsgCallback;
import org.nodex.mods.amqp.AmqpProps;
import org.nodex.mods.amqp.Channel;
import org.nodex.mods.amqp.ChannelHandler;
import org.nodex.mods.amqp.ChannelPool;
import org.nodex.mods.redis.RedisClient;
import org.nodex.mods.redis.RedisConnectHandler;
import org.nodex.mods.redis.RedisConnection;
import org.nodex.mods.redis.ResultHandler;
import org.nodex.mods.stomp.StompClient;
import org.nodex.mods.stomp.StompConnectHandler;
import org.nodex.mods.stomp.StompConnection;
import org.nodex.mods.stomp.StompMsgCallback;
import org.nodex.mods.stomp.StompServer;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

public class CompositionExample {
  public static void main(String[] args) throws Exception {
    new CompositionExample().run();

    System.out.println("Any key to exit");
    System.in.read();
  }

  private CompositionExample() {
  }

  private static final String AMQP_QUEUE = "availability";
  private static final String STOMP_DESTINATION = "availability-request";

  private void run() {
    stompServer();
    httpServer();
    amqpWorker();
    stompWorker();
  }

  private void stompServer() {
    StompServer.createServer().listen(8181);
  }

  private void httpServer() {
    final ChannelPool chPool = ChannelPool.createPool();
    new HttpServer().requestHandler(new EventHandler<HttpServerRequest>() {
      public void onEvent(final HttpServerRequest req) {
        System.out.println("Request uri is " + req.uri);
        if (req.uri.equals("/")) {
          System.out.println("Serving index page");
          //Serve the main page
          FileSystem.instance.readFile("index.html", new CompletionHandler<Buffer>() {
            public void onCompletion(Buffer data) {
              req.response.write(data);
              req.response.end();
            }

            public void onException(Exception e) {
            }
          });
        } else if (req.uri.startsWith("/submit")) {
          //We have received a requestHandler for price/stock information, so we send a message to Rabbit with the name of the item
          final String item = req.getParam("item");
          chPool.getChannel(new ChannelHandler() {
            public void onCreate(final Channel ch) {
              AmqpProps props = new AmqpProps();
              props.headers.put("item", item);
              ch.request("", AMQP_QUEUE, props, (String) null, new AmqpMsgCallback() {
                public void onMessage(AmqpProps respProps, byte[] bod) {
                  //We get a response back with the price and number of items in stock
                  int price = (Integer) respProps.headers.get("price");
                  int stock = (Integer) respProps.headers.get("stock");
                  String content = "<html><body>Price is: " + price + "<br>Stock is: " + stock + "</body></html>";
                   req.response.write(content, "UTF-8").end();
                }
              });
            }
          });
        }
      }
    }).listen(8080);
  }

  private void amqpWorker() {

    //First we need to set up the redis and Stomp connections (and add some reference dataHandler)

    final AtomicReference<RedisConnection> redisConn = new AtomicReference<RedisConnection>();
    final Composable redisConnected = new Composable();

    //Create redis connection
    RedisClient.createClient().connect(6379, "localhost", new RedisConnectHandler() {
      public void onConnect(final RedisConnection conn) {
        //We need to add a little reference dataHandler for this prices
        conn.set("bicycle", "125", new Runnable() {
          public void run() {
            conn.set("aardvark", "333", new Runnable() {
              public void run() {
                redisConn.set(conn);
                redisConnected.complete();
              }
            });
          }
        });
      }
    });

    final AtomicReference<StompConnection> stompConn = new AtomicReference<StompConnection>();
    final Composable stompConnected = new Composable();

    //Create STOMP connection
    StompClient.connect(8181, new StompConnectHandler() {
      public void onConnect(StompConnection conn) {
        stompConn.set(conn);
        stompConnected.complete();
      }
    });

    // Once the connections are setup (asynchronously) we can start the worker
    new Composer().when(redisConnected, stompConnected).
        when(new Runnable() {
          public void run() {
            setupConnections(redisConn.get(), stompConn.get());
          }
        }).end();
  }

  private void setupConnections(final RedisConnection redisConn, final StompConnection stompConn) {

    // Create and start the worker
    AmqpClient.createClient().connect(new AmqpConnectHandler() {
      public void onConnect(AmqpConnection conn) {
        conn.createChannel(new ChannelHandler() {
          public void onCreate(final Channel ch) {
            //Declare the queue
            ch.declareQueue(AMQP_QUEUE, false, true, true, new Runnable() {
              public void run() {
                ch.subscribe(AMQP_QUEUE, true, new AmqpMsgCallback() {
                  public void onMessage(final AmqpProps props, byte[] body) {
                    final String item = props.headers.get("item").toString();
                    Composer comp = new Composer();

                    // Get price from redis
                    final AtomicInteger price = new AtomicInteger(0);
                    Composable redisGet = redisConn.get(item, new ResultHandler() {
                      public void onResult(String value) {
                        price.set(Integer.parseInt(value));
                      }
                    });

                    Map<String, String> headers = new HashMap<String, String>();
                    headers.put("item", item);

                    // Get stock from STOMP worker
                    final AtomicInteger stock = new AtomicInteger(0);
                    Composable responseReturned = stompConn.request(STOMP_DESTINATION, headers, null, new StompMsgCallback() {
                      public void onMessage(Map<String, String> headers, Buffer body) {
                        int st = Integer.valueOf(headers.get("stock"));
                        stock.set(st);
                      }
                    });

                    comp.when(redisGet, responseReturned)         // Execute redis get and stomp requestHandler/response in parallel
                        .when(new Runnable() {                 // Then send back a response with the price and stock
                          public void run() {
                            props.headers.put("price", price.get());
                            props.headers.put("stock", stock.get());
                            ch.publish("", props.replyTo, props, (byte[]) null);
                          }
                        })
                        .end();
                  }
                });
              }
            });
          }
        });
      }
    });
  }

  /*
  The STOMP worker consumes from the price queue and sends back the number of items in stock for the item
   */
  private void stompWorker() {
    StompClient.connect(8181, new StompConnectHandler() {
      public void onConnect(final StompConnection conn) {
        conn.subscribe(STOMP_DESTINATION, new StompMsgCallback() {
          public void onMessage(Map<String, String> headers, Buffer body) {
            System.out.println("Sending back number of items in stock for item " + headers.get("item"));
            headers.put("stock", String.valueOf((int) (10 * Math.random())));
            conn.send(headers.get("reply-to"), headers, null);
          }
        });
      }
    });
  }


}

File: src/examples/ruby/composition/composition_example.rb
# Copyright 2002-2011 the original author or authors.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not use
# this file except in compliance with the License. You may obtain a copy of the
# License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed
# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
# CONDITIONS OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the License.

require "net"
require "buffer"
require "stomp"
require "http"
require "file_system"
require "amqp"
require "redis"
require "composition"
include Net
include Http
include FileSystem
include Amqp
include Redis
include Stomp
include Composition

# This somewhat elaborate example shows an interaction including HTTP, STOMP, AMQP and Redis
#
# Summary: We have a website which allows the user to check the price and stock count for inventory items.
#
# It contains the following components (In real life these would probably be running on different nodes).
#
# 1) HTTP server. We create an HTTP server which serves the index.html page from disk, and responds to HTTP requests for an item.
# when it receives a request it uses the request-response pattern to send that request in an AMQP message to a queue
# and sets a handler for the response.
# When the response returns it formats the price and stock count in the html page returned to the browser.
#
# 2) AMQP consumer. We create an AMQP consumer that consumes from the AMQP queue, and then does two things in parallel
# a) Call redis to get the price for the item
# b) Send a STOMP message, using the request-response pattern to a STOMP destination to request the stock count
# When both a) and b) asynchronously complete, we return an AMQP message as the response to the HTTP server that made
# the request
#
# 3) A redis server
#
# 4) A STOMP server
#
# 5) STOMP consumer. We create a STOMP consumer that subscribes to the STOMP destination, calculates a stock count and
# sends that back in a response message

AMQP_QUEUE = "availability"
STOMP_DESTINATION = "availability-request"

def http_server
  channel_pool = ChannelPool.create_pool
  HttpServer.create_server { |conn|
    conn.request { |req, resp|
      puts "Request uri is #{req.uri}"
      if req.uri == "/"
        puts "Serving index page"
        # Serve the main page
        FileSystem.read_file("index.html") { |data|
          resp.write_buffer(data)
          resp.end
        }
      elsif req.uri.start_with? "/submit"
        # Handle the request for the item
        item = req.get_param("item")
        channel_pool.get_channel { |chan|
          props = Props.new
          props.headers["item"] = item
          chan.request("", AMQP_QUEUE, props, nil) { |resp_props, body|
          # We get a response back with the price and number of items in stock
            price = resp_props.headers["price"]
            stock = resp_props.headers["stock"]
            content = "<html><body>Price is: #{price}<br>Stock is: #{stock}</body></html>"
            resp.write_str(content, "UTF-8").end
          }
        }
      end
    }
  }.listen(8080, "localhost")
end

def amqp_worker

  #First we need to setup the connections and add some reference data

  # First we need to create a connection to redis
  redis_conn = nil
  redis_connected = Completion.create
  RedisClient.create_client.connect(6379, "localhost") { |conn|
  # We add a little reference data that we're going to need later
    conn.set("bicycle", "125") {
      conn.set("aardvark", "333") {
        redis_conn = conn
        redis_connected.complete
      }
    }
  }

  # And we create a connection to the STOMP broker
  stomp_conn = nil
  stomp_connected = Completion.create
  StompClient.connect(8181) { |conn|
    stomp_conn = conn
    stomp_connected.complete
  }

  # We need to make sure we are connected before we can do anything
  Composer.compose.when(redis_connected, stomp_connected).when { do_amqp_worker(redis_conn, stomp_conn) }.end
end

def do_amqp_worker(redis_conn, stomp_conn)

  # Create and start the AMQP worker

  AmqpClient.create_client.connect { |conn|
    conn.create_channel { |chan|
      chan.declare_queue(AMQP_QUEUE, false, true, true) {
        chan.subscribe(AMQP_QUEUE, true) { |props, body|
          item = props.headers["item"].to_s
          comp = Composer.compose

          # Get the price from redis
          price = nil
          redis_get = redis_conn.get(item) { |val|
            price = val
          }

          # Get the stock from the STOMP worker (request-response pattern)
          stock = nil
          response_returned = stomp_conn.request(STOMP_DESTINATION, {"item" => item}, nil) { |headers, body|
            stock = headers["stock"]
          }

          comp.when(redis_get, response_returned).# Get price and stock information
              when { props.headers["price"] = price # Format response message with info before sending
          props.headers["stock"] = stock
          chan.publish_with_props("", props.reply_to, props, nil) }.end
        }
      }
    }
  }
end

def stomp_worker

  # The STOMP worker consumes from the price queue and sends back the number of items in stock for the item

  StompClient.connect(8181) { |conn|
    conn.subscribe(STOMP_DESTINATION) { |headers, body|
      puts "Sending back number of items in stock for item #{headers["item"]}"
      headers["stock"] = rand(10).to_s
      conn.send_with_headers(headers["reply-to"], headers, nil)
    }
  }
end

puts "Starting composition example"

http_server()
amqp_worker()
stomp_worker()

STDIN.gets
